Raft pre elections should be compatible with refuse_to_be_leader

andrewkerr9000 opened the pull request.

andrewkerr9000 committed:
Raft pre elections should be compatible with refuse_to_be_leader

andrewkerr9000 commented:
I'll rearrange code a bit to make the diff clearer and get rid of the TODOs. Also fix committer name.

andrewkerr9000 committed:
Update term if receive vote request from later term for all roles.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java:
```diff
@@ -167,49 +192,127 @@ public Outcome handle( RaftMessages.NewEntry.BatchRequest batchRequest ) throws
+        @Override
+        public Outcome handle( RaftMessages.Timeout.Election election, Outcome outcome, ReadableRaftState ctx, Log log ) throws IOException
+        {
+            log.info( "Election timeout triggered but refusing to be leader" );
```
This will be very useful in debugging!

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java:
```diff
@@ -167,49 +192,127 @@ public Outcome handle( RaftMessages.NewEntry.BatchRequest batchRequest ) throws
+            Set<MemberId> memberIds = ctx.votingMembers();
+            if ( memberIds != null && memberIds.contains( ctx.myself() ) )
+            {
+                outcome.setPreElection( true );
```
Why are we manipulating this if we are anyway refusing to be leader?

andrewkerr9000 replied:
`preElection` needs to be true to participate in pre elections (`PreVoteRequestVotingHandler`). However it won't send pre vote requests and it will ignore pre vote responses.

See lines 361-376.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java:
```diff
@@ -81,17 +81,24 @@ public Outcome handle( RaftMessages.RaftMessage message, ReadableRaftState ctx,
             this.ctx = ctx;
             this.log = log;
             this.outcome = new Outcome( FOLLOWER, ctx );
+            this.preVoteRequestHandlerDelegate = preVoteRequestHandlerDelegate;
```
Is there a reason to single this one out with Delegate in the name? Hard for me to see without IDE-support, but can't really notice anything special through cursory browsing here.

andrewkerr9000 replied:
Probably an oversight from an earlier version.

martinfurmanski commented:
I had just very minor comments. I think the functionality looks good enough for a merge, however the PR run seems to have failed miserably. Don't know if just flaky or perhaps some lifecycle issue in test refactoring.

andrewkerr9000 commented:
Looks like a PreElectionIT is failing, and causing a cascade of failures by not cleaning up after itself.

andrewkerr9000 committed:
SharedDiscoveryService works with refuseToBeLeader

Topologies are bootstrapped by the first server that does not refuse to be leader. The bootstrapping member must become leader to upload the snapshot to the other members.

This logic also exists in HazelcastClusterTopology, but is coupled with the underlying data structures in a way that is hard to extract.

andrewkerr9000 commented:
@martinfurmanski Tests should be stable now.

tl;dr: we need to prevent members that refuse to be leader bootstrapping the cluster. HazelcastDiscoveryService already did this, but SharedDiscoveryService did not. Unfortunately the logic is coupled to implementation specific data structures and cannot easily be extracted further up the call hierarchy.

andrewkerr9000 commented:
Cypher test failed, rebased.
